src/Polynomials/MonomialBases.jl

```diff
@@ -406,6 +408,34 @@ function _gradient_nd!(
 
 end
 
+function _gradient_nd!(
+  v::AbstractVector{G},
+  x,
+  orders,
+  terms::AbstractVector{CartesianIndex{D}},
+  c::AbstractMatrix{T},
+  g::AbstractMatrix{T},
+  ::Type{V},
+  ::Val{true}) where {G,T,D,V}
+
+  z = zero(Mutable(VectorValue{D,T}))
+  _gradient_nd!(v,x,orders,terms,c,g,z,T)
+end
+
+function _gradient_nd!(
+  v::AbstractVector{G},
+  x,
+  orders,
+  terms::AbstractVector{CartesianIndex{D}},
+  c::AbstractMatrix{T},
+  g::AbstractMatrix{T},
+  ::Type{V},
+  ::Val{false}) where {G,T,D,V}
+
+  z = zeros(T,D)
```

Comment:
you still have the dynamic memory allocation.  I would move to the ancestors, possibly into the cache of the kernel, and pass it to this function as an extra parameter.

Reply:
ok, do you mean to have 2 functions:
```julia
return_cache(
  fg::FieldGradientArray{1,MonomialBasis{D,V}},
  x::AbstractVector{<:Point},
  TisbitsType::Val{true}) 
```
and 
```julia
return_cache(
  fg::FieldGradientArray{1,MonomialBasis{D,V}},
  x::AbstractVector{<:Point},
  TisbitsType::Val{false}) 
```
?

Reply:
Yes. This is more or less the idea. However, note that the interface of return_cache and evaluate! has to fit into the LazyArray umbrella. To this end, we cannot add the extra parameter to return_cache and evaluate! as you propose. 
Instead, we can define private functions (note the _ prefix) 

```julia
_return_cache(
  fg::FieldGradientArray{1,MonomialBasis{D,V}},
  x::AbstractVector{<:Point},
  TisbitsType::Val{true}) 
```

```julia
_return_cache(
  fg::FieldGradientArray{1,MonomialBasis{D,V}},
  x::AbstractVector{<:Point},
  TisbitsType::Val{false}) 
```

return_cache detects whether TisbitsType is true and false, and calls the appropiate _return_cache variant. The same idea for evaluate!. Do u understand what I mean?

Reply:
Hey @amartinhuertas, I have made a commit. Please see if this is better.

Reply:
Thanks @ConnorMallon ! 

I have already taken a look at it. We are almost there in regards to the final solution to this particular issue.

I still have some minor comments, that I will add straight away now along with the code. Once these are taken into account, I would resolve this conversation.

Reply:
I have already added my comments.